src/write.c

```diff
@@ -1219,10 +1224,173 @@ static avifResult avifEncoderAddImageInternal(avifEncoder * encoder,
+
+    avifResult result = avifEncoderAddImageInternal(encoder, gridCols, gridRows, cellImages, durationInTimescales, addImageFlags);
+    if (result == AVIF_RESULT_OK) {
+        result = avifEncoderFinish(encoder, output);
+    }
+    *ioStats = encoder->ioStats;
+    *diag = encoder->diag;
+    avifEncoderDestroy(encoder);
+    return result;
+}
+
+static avifResult avifEncoderAddImageInternalTargetSize(avifEncoder * encoder,
+                                                        uint32_t gridCols,
+                                                        uint32_t gridRows,
+                                                        const avifImage * const * cellImages,
+                                                        uint64_t durationInTimescales,
+                                                        avifAddImageFlags addImageFlags)
+{
+    if (encoder->targetSize == 0) {
+        return avifEncoderAddImageInternal(encoder, gridCols, gridRows, cellImages, durationInTimescales, addImageFlags);
+    }
+
+    // Setting avifEncoder.targetSize for one call to avifEncoderAddImage() without knowing if there will
+    // be another frame is ambiguous. The user may expect the whole animation to match the targetSize, or
+    // may expect that it applies to each frame. Allowing avifEncoder.targetSize>0 for animations would also
+    // make the code more complicated, because all input frames would be stored until avifEncoderFinish().
+    // For simplicity, it is reserved to single-frame images.
+    AVIF_CHECKERR(addImageFlags & AVIF_ADD_IMAGE_FLAG_SINGLE, AVIF_RESULT_INVALID_ARGUMENT);
+
+    if ((encoder->quality != AVIF_QUALITY_DEFAULT) && (encoder->qualityAlpha != AVIF_QUALITY_DEFAULT)) {
+        // If both quality and qualityAlpha are forced to a given value (not AVIF_QUALITY_DEFAULT),
+        // there is no binary search to do.
+        return avifEncoderAddImageInternal(encoder, gridCols, gridRows, cellImages, durationInTimescales, addImageFlags);
+    }
+
+    // avifCodecEncodeImageFunc() uses a quantizer argument instead of the avifEncoder.quality one.
+    // The quantizer range [0:63] being smaller than the quality range [0:100], it would be faster
+    // to operate the binary search below directly on the quantizer setting rather than on the quality one.
+    // For simplicity, it uses the quality for now.
+
+    // Use the encoder instance as a "current settings" holder.
+    const avifBool searchForQuality = encoder->quality == AVIF_QUALITY_DEFAULT;
+    const avifBool searchForQualityAlpha = encoder->qualityAlpha == AVIF_QUALITY_DEFAULT;
+
+    int closestQuality = AVIF_QUALITY_DEFAULT;
+    avifRWData closestEncoded = AVIF_DATA_EMPTY;
+    size_t closestSizeDiff = 0;
+    avifIOStats closestIoStats = { 0, 0 };
+
+    int minQuality = AVIF_QUALITY_WORST; // inclusive
+    int maxQuality = AVIF_QUALITY_BEST;  // inclusive
+    while (minQuality <= maxQuality) {
+        const int quality = (minQuality + maxQuality) / 2;
```

Comment:
webp uses a different approach since it works with a lower number of passes [1]. For both I'm wondering if we could make an early guess based on the image resolution and target size to cut down on the iterations, but it's better to start with this approach.

[1]: https://chromium.googlesource.com/webm/libwebp/+/refs/tags/v1.3.0/src/enc/frame_enc.c#71